Document where clauses in record fields

Villetaneuse opened the pull request:
<!-- Thank you for your contribution.
     Make sure you read the contributing guide and fill this template. -->


<!-- If this is a bug fix, make sure the bug was reported beforehand. -->
Fixes / closes #20731

- [x] Added / updated **documentation**.
- [x] Updated **documented syntax** by running `make doc_gram_rsts`.

Rendered: https://coq.gitlabpages.inria.fr/-/coq/-/jobs/5866597/artifacts/_build/default/doc/refman-html/language/core/records.html

jfehrle commented on doc/sphinx/language/core/records.rst:
```diff
@@ -177,6 +177,37 @@ Defining record types
+           cond: forall a b, a & b = b & a
+         }.
+
+      However, it is not possible to add a :g:`where` clause for the record itself:
+
+      .. rocqtop:: all
+
+         Reserved Notation "a & b" (at level 40).
+         Fail Record foo (op : nat -> nat -> nat) := mkfoo
+         {
+           point: nat;
+           cond: forall a b, a & b = b & a
+         } where "a & b" := (op a b).
+
+   .. exn:: Error: "where" clause not supported for records.
+      :undocumented:
+
```
Including examples of syntax that the documented grammar clearly doesn't support seems overly verbose.  I would drop this part (keep the `.. exn` part if you like).  There's an infinity of inputs that the grammar doesn't allow--why is this one so special?

SkySkimmer replied:
the grammar does allow it, which is why it produces a regular error and not a syntax error

jfehrle replied:
Not much point in describing syntax that always gets an error.  The parser parses it, but the edited grammar shown in this section does not (intentionally).  For which of these commands are decl_notations allowed?  Looks like only for `Inductive` and `CoInductive`.

```

inductive_or_record_definition: [
| opt_coercion cumul_ident_decl binders OPT [ "|" binders ] OPT [ ":" lconstr ] opt_constructors_or_fields decl_notations
]

gallina: [
| finite_token inductive_or_record_definition
| inductive_token LIST1 inductive_or_record_definition SEP "with"
]

finite_token: [
| "Variant"
| "Record"
| "Structure"
| "Class"
]

inductive_token: [
| "Inductive"
| "CoInductive"
]
```

Villetaneuse replied:
It's probably the first question the user will ask. I think it's better to document it.
And, the error was not documented before. It's better if it comes with an example.

jfehrle replied:
What's the question the user will ask?  The example doesn't address why it's an error.  The error message alone seems clear and it's also very clear from the syntax given in the document.  The example doesn't add anything to it.  It's very easy to describe the limitation in the command description.  Adding very obvious examples is not a service to the reader, particularly when they only show how to generate error messages--it's just more verbiage to wade through.

Villetaneuse replied:
I removed the failing example.

jfehrle commented on doc/sphinx/language/core/records.rst:
```diff
@@ -177,6 +177,37 @@ Defining record types
+
+      Records themselves do not support a :g:`where` clause, but record fields do,
+      with a somewhat unintuitive behaviour: the notation defined in the where clause
+      is active after the field where it appears, not in the field itself.
+
+      .. rocqtop:: all
+
+         Reserved Notation "a & b" (at level 40).
+         Record foo (op : nat -> nat -> nat) := mkfoo
+         {
+           point: nat where "a & b" := (op a b);
+           cond: forall a b, a & b = b & a
+         }.
```
How is a record field notation different from non-record field notations?  Does it behave differently?  Why is it useful?  Can you include info that somewhere?  It would helpful to add a line that uses the record field notation.

Also, it would be better if you can come up with a realistic example.  This seems contrived.  I don't understand how `a & b` relates to point, which is merely a nat.

Villetaneuse replied:
I can't because the feature is broken. It has to be contrived since the behaviour of "where clause" is (probably by chance) weird in records. If you want I can add `op` as a field in the Record, but the point that it's not effective in the field where the `where` clause is written must be clear.
It will probably be deprecated in the near future, but we have to document the present.

SkySkimmer replied:
see  https://rocq-prover.zulipchat.com/#narrow/channel/423352-Rocq-Notation-system/topic/.E2.9C.94.20Reserved.20Notation.20for.20custom.20scope for an example

jfehrle reviewed (commented):
Would you improve this?

Villetaneuse commented:
To address your remarks:
- moved and reworked explanation outside the example
- changed the example to a more realistic use case
- removed the failing example
- document (shortly) the error when "where" clause is used at Record level

Villetaneuse commented:
Thank you for your remarks.
Doing changes now. However I do want to keep the emphasis on the fact that these record field clauses do not apply to the current field (which is not intuitive).

jfehrle commented on doc/sphinx/language/core/records.rst:
```diff
@@ -113,6 +113,13 @@ Defining record types
+     defines notations which are active after this field, not in the field
+     itself and only until the end of the :cmd:`Record`
+     (see :ref:`example <record_where_clause>`).
+     Note that :g:`where` clauses cannot be added at the record level.
```
Ah, would you also move this text to just after line 89 (reread the first paragraph in my previous comment)?

Villetaneuse replied:
Sorry, I missed that.

SkySkimmer commented on doc/sphinx/language/core/records.rst:
```diff
@@ -177,6 +184,26 @@ Defining record types
+
+   .. exn:: Error: "where" clause not supported for records.
+
+      A :g:`where` clause appears at the :cmd:`Record` level.
```
The error is not based on the command name, eg `Inductive foo := { x : nat } where ...` will error

SkySkimmer replied:
It errors for records, including records defined with the Inductive keyword

jfehrle commented on doc/sphinx/language/core/records.rst:
```diff
@@ -177,6 +183,27 @@ Defining record types
+
+   .. exn:: Error: "where" clause not supported for records.
+
+      :g:`where` clauses are only supported for :n:`@record_field`\s, not for the overall
+      :cmd:`Record` definition.
```
This error message can appear for Record, Structure, Class, Inductive and CoInductive, so not quite right to refer to the Record command here.  But they are all definitions.

```suggestion
      :g:`where` clauses are only supported for :n:`@record_field`\s, not for the overall
      definition.
```

FWIW, the message wording `Error: "where" clause not supported for records.` is not great--Inductive and CoInductive are not records at all.  Class is a special case of a record.  Not sure about Structure--the doc says it's a synonym for Record, but a superficial examination of the code suggests that it's processed differently from Record--do either of you know how it differs?

Villetaneuse replied:
I think we are diverging from the original intent of this PR.
I guess records are not that much about the keyword defining them but about the syntax `record_definition` which makes them records.

jfehrle replied:
> I think we are diverging from the original intent of this PR.

How so?  You wanted to add the exception.  Immediately above, I'm suggesting a tiny correction to what you currently have for that.  And [here](https://github.com/rocq-prover/rocq/pull/20753#discussion_r2150846923) asks you to add your text in a different spot.  Everything else has been covered.

jfehrle commented on doc/sphinx/language/core/records.rst:
```diff
@@ -100,15 +115,6 @@ Defining record types
-     - :n:`{+ @binder } @of_type_inst := @term` is equivalent to
-       :n:`: forall {+ @binder } , @of_type_inst := fun {+ @binder } => @term`
-
      :n:`:= @term`, if present, gives the value of the field, which may depend
      on the fields that appear before it.  Since their values are already defined,
      such fields cannot be specified when constructing a record.
```
Ah, this paragraph should be right after the bullet points (move to line 104).

jfehrle reviewed (approved):
Looks good.  One last tweak.
